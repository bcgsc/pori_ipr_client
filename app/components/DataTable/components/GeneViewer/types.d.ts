import {
  CopyNumberType,
  ExpOutliersType,
  KbMatchType,
  SmallMutationType,
  StructuralVariantType,
} from '@/common';
import { ImageType } from '@/components/Image';

type GeneViewerType = {
  copyNumber: CopyNumberType[];
  expDensityGraph: ImageType[];
  expRNA: ExpOutliersType[];
  kbMatches: KbMatchType[];
  smallMutations: SmallMutationType[];
  structuralVariants: StructuralVariantType[];
};

export {
  GeneViewerType,
};
