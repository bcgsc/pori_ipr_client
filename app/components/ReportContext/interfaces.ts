interface PatientInformationInterface {
  age: string | null,
  biopsySite: string | null,
  caseType: 'Pediatric' | 'Adult',
  constitutionalProtocol: string | null,
  constitutionalSample: string | null,
  createdAt: string,
  diagnosis: string | null,
  gender: string | null,
  ident: string,
  physician: string | null,
  reportDate: string | null,
  tumourProtocol: string | null,
  tumourSample: string | null,
  updatedAt: string | null,
}

interface SampleInfoInterface {
  biopsySite: string | null,
  collectionDate: string | null,
  'Patho TC': string | null,
  'Primary Site': string | null,
  Sample: string | null,
  'Sample Name': string | null,
}

interface ReportContextInterface {
  /** Current report that's being viewed */
  report: {
    alternateIdentifier?: string,
    biopsyName?: string,
    createdAt: string,
    ident: string,
    patientId: string,
    patientInformation: PatientInformationInterface,
    ploidy: string,
    sampleInfo: Array<SampleInfoInterface>
    state: string,
    subtyping: string,
    tumourContent?: number,
    type: string,
  } | null,
  /** Set new current report */
  setReport: (newValue: Record<string, unknown>) => void,
}

export {
  PatientInformationInterface,
  ReportContextInterface,
};
