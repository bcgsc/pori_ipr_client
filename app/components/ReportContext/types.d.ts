import { RecordDefaults } from '@/common';

type PatientInformationType = {
  age: string | null;
  biopsySite: string | null;
  caseType: 'Pediatric' | 'Adult';
  constitutionalProtocol: string | null;
  constitutionalSample: string | null;
  diagnosis: string | null;
  gender: string | null;
  physician: string | null;
  reportDate: string | null;
  tumourProtocol: string | null;
  tumourSample: string | null;
} & RecordDefaults;

type SampleInfoType = {
  biopsySite: string | null;
  collectionDate: string | null;
  'Patho TC': string | null;
  'Primary Site': string | null;
  Sample: string | null;
  'Sample Name': string | null;
};

type ReportType = {
  alternateIdentifier?: string;
  analysisStartedAt: string | null;
  biopsyName?: string;
  kbVersion: string;
  patientId: string;
  patientInformation: PatientInformationType,
  ploidy: string;
  reportVersion: string;
  sampleInfo: SampleInfoType[];
  state: string;
  subtyping: string;
  tumourContent?: number;
  type: string;
} & RecordDefaults;

type ReportContextType = {
  /** Current report that's being viewed */
  report: ReportType | null,
  /** Set new current report */
  setReport: (newReport: ReportType) => void;
};

export {
  PatientInformationType,
  ReportContextType,
  ReportType,
  SampleInfoType,
};
