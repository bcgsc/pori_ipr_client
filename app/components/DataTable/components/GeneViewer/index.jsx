import React, { useEffect, useState } from 'react';
import PropTypes from 'prop-types';
import {
  Dialog,
  DialogTitle,
  DialogContent,
  DialogActions,
  Typography,
  IconButton,
  Button,
} from '@material-ui/core';
import { HighlightOff } from '@material-ui/icons';
import DataTable from '../..';
import geneViewerApi from '../../../../services/reports/geneViewer';

/**
 * @param {object} props props
 * @param {func} props.onClose parent close handler
 * @param {object} props.selectedRow current row object
 * @param {bool} props.open is open?
 * @return {*} JSX
 */
function GeneViewer(props) {
  const {
    onClose,
    open,
    gene,
    reportId,
  } = props;
  
  const [geneData, setGeneData] = useState();

  const handleClose = (value) => {
    onClose(value);
  };

  useEffect(() => {
    if (open) {
      const api = async () => {
        const resp = await geneViewerApi(gene, reportId);
        setGeneData(resp);
      };
      api();
    }
  }, [open]);

  return (
    <Dialog
      onClose={handleClose}
      open={open}
      maxWidth="xl"
      fullWidth
    >
      <DialogTitle>
        <span className="dialog__title">
          <Typography variant="h6" align="center">
            Gene Viewer
          </Typography>
          <IconButton onClick={handleClose}>
            <HighlightOff />
          </IconButton>
        </span>
      </DialogTitle>
      <DialogContent>
        test
      </DialogContent>
      <DialogActions>
        <Button color="primary" onClick={handleClose}>
          Close
        </Button>
      </DialogActions>
    </Dialog>
  );
}

GeneViewer.propTypes = {
  onClose: PropTypes.func.isRequired,
  gene: PropTypes.any.isRequired,
  reportId: PropTypes.string.isRequired,
  open: PropTypes.bool.isRequired,
};

export default GeneViewer;
