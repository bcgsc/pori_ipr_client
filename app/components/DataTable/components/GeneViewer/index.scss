.dialog {
  &__title {
    display: flex;
    justify-content: space-between;
    align-items: center;
  }
}

.tab {
  &--center {
    text-align: center;
    margin: 12px 0 0;
  }
}