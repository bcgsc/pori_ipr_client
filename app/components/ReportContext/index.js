import React from 'react';

const ReportContext = React.createContext({
  report: null,
  setReport: () => {},
});

export default ReportContext;
