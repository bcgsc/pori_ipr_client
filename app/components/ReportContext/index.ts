import React from 'react';
import { ReportContextType } from './types';

const ReportContext = React.createContext<ReportContextType>({
  report: null,
  setReport: () => {},
});

export default ReportContext;
