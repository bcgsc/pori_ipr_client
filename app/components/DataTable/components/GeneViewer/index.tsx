import React, {
  useEffect, useState, useContext, useCallback,
} from 'react';
import {
  Dialog,
  DialogTitle,
  DialogContent,
  DialogActions,
  Typography,
  Button,
  Tabs,
  Tab,
  AppBar,
} from '@mui/material';
import { AgGridReact } from '@ag-grid-community/react';

import api from '@/services/api';
import snackbar from '@/services/SnackbarUtils';
import ReportContext from '@/context/ReportContext';
import { columnDefs } from '@/views/ReportView/components/KbMatches/columnDefs';
import { columnDefs as smallMutationsColumnDefs } from '@/views/ReportView/components/SmallMutations/columnDefs';
import copyNumberColumnDefs from '@/views/ReportView/components/CopyNumber/columnDefs';
import expressionColumnDefs from '@/views/ReportView/components/Expression/columnDefs';
import structuralVariantsColumnDefs from '@/views/ReportView/components/StructuralVariants/columnDefs';
import { GeneViewerType } from './types';

import KbMatchesActionCellRenderer from '../KbMatchesActionCellRenderer';
import { ActionCellRenderer } from '../ActionCellRenderer';
import NoRowsOverlay from '../NoRowsOverlay';

import './index.scss';

const defaultTableOptions = {
  frameworkComponents: {
    ActionCellRenderer,
    KbMatchesActionCellRenderer,
    NoRowsOverlay,
  },
  noRowsOverlayComponent: 'NoRowsOverlay',
  enableCellTextSelection: true,
};

type GeneViewerProps = {
  isLink: boolean;
  gene: string;
};

const GeneViewer = ({
  gene,
  isLink = false,
}: GeneViewerProps): JSX.Element => {
  const { report } = useContext(ReportContext);

  const [isOpen, setIsOpen] = useState(false);
  const [geneData, setGeneData] = useState<GeneViewerType>();
  const [tabValue, setTabValue] = useState(0);

  const handleTabChange = useCallback((_event, newValue: number) => {
    setTabValue(newValue);
  }, []);

  useEffect(() => {
    if (isOpen && gene) {
      const getData = async () => {
        try {
          const resp = await api.get(`/reports/${report.ident}/gene-viewer/${gene}`).request();
          setGeneData(resp);
        } catch {
          snackbar.error(`Error: gene viewer data does not exist for ${gene}`);
          setIsOpen(false);
        }
      };
      getData();
    }
  }, [gene, isOpen, report]);

  if (!gene) return null;

  if (!isLink) {
    return <span>{gene}</span>;
  }

  return (
    <>
      <span
        tabIndex={0}
        role="button"
        onClick={() => setIsOpen(true)}
        onKeyDown={() => setIsOpen(true)}
        className="gene__text"
      >
        {gene}
      </span>
      <Dialog
        onClose={() => setIsOpen(false)}
        open={isOpen}
        maxWidth="xl"
        fullWidth
      >
        <DialogTitle>
          <span className="dialog__title">
            <Typography variant="h5" align="center">
              Gene Viewer
            </Typography>
          </span>
        </DialogTitle>
        <DialogContent>
          <AppBar position="static" elevation={0} color="transparent">
            <Tabs
              value={tabValue}
              onChange={handleTabChange}
              variant="scrollable"
              scrollButtons="auto"
            >
              {geneData && Object.entries(geneData).map(([key, value]) => (
                <Tab key={key} label={`${key} (${value.length})`} />
              ))}
            </Tabs>
          </AppBar>
          {geneData && (
          <>
            {tabValue === 0 && (
              <div className="ag-theme-material">
                <AgGridReact
                  rowData={geneData.kbMatches}
                  columnDefs={columnDefs}
                  domLayout="autoHeight"
                  {...defaultTableOptions}
                />
              </div>
            )}
            {tabValue === 1 && (
              <div className="ag-theme-material">
                <AgGridReact
                  rowData={geneData.smallMutations}
                  columnDefs={smallMutationsColumnDefs}
                  domLayout="autoHeight"
                  {...defaultTableOptions}
                />
              </div>
            )}
            {tabValue === 2 && (
              <div className="ag-theme-material">
                <AgGridReact
                  rowData={geneData.copyNumber}
                  columnDefs={copyNumberColumnDefs}
                  domLayout="autoHeight"
                  {...defaultTableOptions}
                />
              </div>
            )}
            {tabValue === 3 && (
              <div className="ag-theme-material">
                <AgGridReact
                  rowData={geneData.expRNA}
                  columnDefs={expressionColumnDefs}
                  domLayout="autoHeight"
                  {...defaultTableOptions}
                />
              </div>
            )}
            {tabValue === 4 && (
            <div className="tab--center">
              {geneData.expDensityGraph.map((graph) => (
                <img
                  key={graph.ident}
                  src={`data:image/png;base64,${graph.data}`}
                  alt="Expression Density Graph"
                />
              ))}
            </div>
            )}
            {tabValue === 5 && (
              <div className="ag-theme-material">
                <AgGridReact
                  rowData={geneData.structuralVariants}
                  columnDefs={structuralVariantsColumnDefs}
                  domLayout="autoHeight"
                  {...defaultTableOptions}
                />
              </div>
            )}
          </>
          )}
        </DialogContent>
        <DialogActions>
          <Button color="primary" onClick={() => setIsOpen(false)}>
            Close
          </Button>
        </DialogActions>
      </Dialog>
    </>
  );
};

export default GeneViewer;
