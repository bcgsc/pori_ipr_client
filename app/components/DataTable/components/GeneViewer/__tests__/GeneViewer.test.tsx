import React from 'react';
import { when, resetAllWhenMocks } from 'jest-when';
import {
  screen, render, waitFor,
} from '@testing-library/react';
import { ModuleRegistry } from '@ag-grid-community/core';
import { ClientSideRowModelModule } from '@ag-grid-community/client-side-row-model';
import ReportContext, { ReportType } from '@/context/ReportContext';
import api, { ApiCall } from '@/services/api';
import GeneViewer from '..';

const mockGene = 'TP53';
const mockErrorGene = 'NonExistent25';
const mockGeneResults = {
  copyNumber: [],
  expDensityGraph: [],
  expRNA: [],
  kbMatches: [],
  smallMutations: [],
  structuralVariants: [],
};

jest.mock('@/services/api');
jest.mock('@/services/SnackbarUtils');

const mockReport = {
  ident: '0557a-78bba-2009c-c470f',
} as ReportType;

const withReportContext = (Component) => function ReportContextHOC(props) {
  return (
    <ReportContext.Provider value={{ report: mockReport, setReport: () => {} }}>
      <Component {...props} />
    </ReportContext.Provider>
  );
};

describe('GeneViewer', () => {
  let Component;
  beforeAll(() => {
    jest.resetModules();
    ModuleRegistry.registerModules([
      ClientSideRowModelModule,
    ]);
  });
  beforeEach(() => {
    Component = withReportContext(GeneViewer);
    resetAllWhenMocks();
  });

  test('It matches the snapshot', () => {
    const { asFragment } = render(
      <Component
        gene={mockGene}
        isOpen
      />,
    );

    expect(asFragment()).toMatchSnapshot();
  });

  test('It calls onClose if the gene does not exist', async () => {
    when(api.get as (endpoint: string) => Partial<ApiCall>)
      .calledWith(`/reports/${mockReport.ident}/gene-viewer/${mockErrorGene}`)
      .mockImplementation(() => ({ request: async () => { throw new Error(); } }));

    const mockOnClose = jest.fn();

    render(
      <Component
        gene={mockErrorGene}
        isOpen
        onClose={mockOnClose}
      />,
    );

    await waitFor(() => expect(mockOnClose).toHaveBeenCalledTimes(1));
    await waitFor(() => expect(mockOnClose).toHaveBeenCalledWith());
  });

  test('Tabs are all shown', async () => {
    when(api.get as (endpoint: string) => Partial<ApiCall>)
      .calledWith(`/reports/${mockReport.ident}/gene-viewer/${mockGene}`)
      .mockImplementation(() => ({ request: async () => mockGeneResults }));

    render(
      <Component
        gene={mockGene}
        isOpen
        onClose={() => {}}
      />,
    );

    const promises = [];
    for (const key of Object.keys(mockGeneResults)) {
      promises.push(screen.findByText(key, { exact: false }));
    }
    const resolved = await Promise.all(promises);
    for (const elem of resolved) {
      expect(elem).toBeInTheDocument();
    }
  });
});
