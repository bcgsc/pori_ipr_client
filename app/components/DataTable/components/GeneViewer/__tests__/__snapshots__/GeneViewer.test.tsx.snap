// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GeneViewer It matches the snapshot 1`] = `<DocumentFragment />`;
